File: Longest Square Streak in an Array.py
class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        nums = list(set(nums))
        n = len(nums)
        terms = {v: i for i, v in enumerate(nums)}
        parents = [i for i in range(n)]
        counts = [1] * n

        def find(i):
            nonlocal parents
            if parents[i] != i:
                return find(parents[i])
            return i

        def union(i, j):
            nonlocal parents, counts
            parent_i = find(i)
            parent_j = find(j)
            if parent_i != parent_j:
                parents[parent_i] = parent_j
                counts[parent_j] += counts[parent_i]
                counts[parent_i] = 0

        for i, num in enumerate(nums):
            nxt = num * num
            if nxt in terms:
                union(i, terms[nxt])

        maxx = max(counts)
        if maxx < 2:
            return -1
        return maxx

# class Solution:
#     def longestSquareStreak(self, nums: List[int]) -> int:
#         dp = set(nums)
#         maxx = -1
#         for num in set(nums):
#             if num not in dp:
#                 continue
#
#             length = 1
#             dp.remove(num)
#
#             left = num
#             while True:
#                 sqrt = int(math.sqrt(left))
#                 root = sqrt ** 2
#                 if root == left and sqrt in dp:
#                     left = sqrt
#                     length += 1
#                     dp.remove(sqrt)
#                 else:
#                     break
#
#             right = num
#             while True:
#                 sq = right ** 2
#                 if sq in dp:
#                     right = sq
#                     length += 1
#                     dp.remove(sq)
#                 else:
#                     break
#
#             if length > 1:
#                 maxx = max(maxx, length)
#         return maxx

# class Solution:
#     def longestSquareStreak(self, nums: List[int]) -> int:
#         nums.sort()
#         dp = {}
#         maxx = -1
#         for num in nums:
#             sqrt = int(math.sqrt(num))
#             root = sqrt ** 2
#             if root == num and sqrt in dp:
#                 dp[num] = dp[sqrt] + 1
#                 maxx = max(maxx, dp[num])
#             else:
#                 dp[num] = 1
#         return maxx

File: Longest Square Streak in an Array.go
func longestSquareStreak(nums []int) int {
    terms := make(map[int]int)
    nums2 := make([]int, 0)
    parents := make([]int, 0)
    counts := make([]int, 0)
    maxx := 1
    j := 0

    for _, num := range nums {
        if _, ok := terms[num]; !ok {
            terms[num] = j
            nums2 = append(nums2, num)
            parents = append(parents, j)
            j++
            counts = append(counts, 1)
        }
    }

    var find func(i int)int
    var union func(i, j int)

    find = func(i int) int {
        if parents[i] != i {
            return find(parents[i])
        }
        return i
    }

    union = func(i, j int) {
        parent_i := find(i)
        parent_j := find(j)
        if parent_i != parent_j {
            parents[parent_i] = parent_j
            counts[parent_j] += counts[parent_i]
            maxx = max(maxx, counts[parent_j])
        }
    }

    for i, num := range nums2 {
        if num >= 317 {
            continue
        }

        nxt := num * num
        if val, ok := terms[nxt]; ok {
            union(i, val)
        }
    }

    if maxx < 2 {
        return -1
    }
    return maxx
}
